3p/ampcontext.js

```diff
@@ -40,18 +53,63 @@ export class AmpContext {
     /** @type {?string} */
     this.sentinel = null;
 
-    // TODO(alanorozco): confirm type
-    /** @type {?} */
+    /** @type {?number} */
     this.startTime = null;
 
-    // TODO(alanorozco): confirm type
     /** @type {?string} */
     this.referrer = null;
 
+    /** @type {?boolean} */
+    this.hidden = null;
+
+    /** @type {?Object} */
+    this.mode = null;
+
+    /** @type {?Object} */
+    this.initialIntersection = null;
+
+    /** @type {?string} */
+    this.sourceUrl = null;
+
+    /** @type {!Observable<Object>} */
+    this.visibilityObservable_ = new Observable();
+
     this.findAndSetMetadata_();
+
+    /** @protected {!IframeMessagingClient} */
     this.client_ = new IframeMessagingClient(win);
     this.client_.setHostWindow(this.getHostWindow_());
     this.client_.setSentinel(dev().assertString(this.sentinel));
+
+    this.listenForPageVisibility_();
```

Comment:
Discussed offline, we will eventually deparate `amp:visibilitychange` event and replace it with observePageVisibility API.

We will keep this in constructor for now, and start a deprecation process once the new API is available.